File: Python/longestSubStr.py
# Given a string s, find the length of the longest substring without duplicate characters.

class LongestSubStr(object):
    def solution(self, s):
        """
        :type s: str
        :rtype: int
        """
        seen = {}
        longest = 0
        start = 0

        for i in range(len(s)):
            c = s[i]
            if c in seen and seen[c] >= start:
                start = seen[c] + 1
            seen[c] = i
            tmp = i - start + 1
            if tmp > longest:
                longest = tmp
        return longest
    
def main():
    lss = LongestSubStr()
    test_str = "abcabcbb"
    result = lss.solution(test_str)
    print(f'Longest substring without duplicates in "{test_str}": {result}')
    
if __name__ == "__main__":
    main()

File: Java/longestSubStr.java
// Given a string s, find the length of the longest substring without duplicate characters.
import java.util.*;

class longestSubStr {
    public int solution(String s) {
        Map<Character, Integer> seen = new HashMap<>();
        int longest = 0;
        int start = 0;

        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if(seen.containsKey(c) && seen.get(c) >= start){
                start = seen.get(c) + 1;
            }

            seen.put(c, i);
            int tmp = i - start + 1;
            if(tmp > longest){longest = tmp;}
        }
        return longest;
    }

    public static void main(String[] args) {
        longestSubStr lss = new longestSubStr();
        String testStr = "abcabcbb";
        int result = lss.solution(testStr);
        System.out.println("Longest substring without duplicates in \"" + testStr + "\": " + result);
    }
}
